server/src/main/java/org/opensearch/cluster/service/MasterTaskThrottler.java

```diff
@@ -0,0 +1,139 @@
+        super(false);
+        clusterSettings.addSettingsUpdateConsumer(ENABLE_MASTER_THROTTLING, this::setMasterThrottlingEnable, this::validateEnableSetting);
+        clusterSettings.addSettingsUpdateConsumer(THRESHOLD_SETTINGS, this::updateSetting, this::validateSetting);
+        this.masterService = masterService;
+    }
+
+    private void setMasterThrottlingEnable(boolean enableMasterThrottling) {
+        super.setThrottlingEnabled(enableMasterThrottling);
+    }
+
+    public void validateEnableSetting(final boolean enableMasterThrottling) {
+        if(masterService.state().nodes().getMinNodeVersion().compareTo(Version.V_7_10_3) < 0) {
+            throw new IllegalArgumentException("All the nodes in cluster should be on version later than 7.10.3");
+        }
+    }
+
+    public void validateSetting(final Settings settings) {
+        Map<String, Settings> groups = settings.getAsGroups();
+        for(String key : groups.keySet()) {
+            int threshold = groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE);
+            if(threshold < DEFAULT_THRESHOLD_VALUE) {
+                throw new IllegalArgumentException("Provide positive integer for limit or -1 for disabling throttling");
+            }
+            if(!taskNameToClassMapping.containsKey(key)) {
+                try {
+                    Class.forName(key.replace("/", "."));
+                } catch (ClassNotFoundException e) {
+                    throw new IllegalArgumentException("Provide valid key, either configured task type or path of class");
+                }
+            }
+        }
+    }
+
+    public void updateSetting(final Settings settings) {
+        Map<String, Settings> groups = settings.getAsGroups();
+        for(String key : groups.keySet()) {
+            if(taskNameToClassMapping.containsKey(key)) {
+                updateLimit(taskNameToClassMapping.get(key), groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE));
+            } else {
+                try {
+                    updateLimit(Class.forName(key.replace("/", ".")),
+                            groups.get(key).getAsInt("value", DEFAULT_THRESHOLD_VALUE));
+                } catch (ClassNotFoundException e) {
+                    // Ideally it should not throw this exception, as we are already validating this in validateSetting.
+                    logger.error("Failed to apply validated master throttler setting, exception: [{}] ", e.getLocalizedMessage());
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean acquire(final Class type, final int permits) {
+        boolean ableToAcquire = super.acquire(type, permits);
+        if(!ableToAcquire) {
+            throttledTasksCount.inc(permits);
+        }
+        return ableToAcquire;
+    }
+
+    public long throttledTaskCount() {
+        return throttledTasksCount.count();
+    }
+
```

Comment:
[Old PR Comment:](https://github.com/opensearch-project/OpenSearch/pull/484#discussion_r611477321)

could we maintain throttledTasksCount at the task level? As I understand this is a cumulative count, wondering should it get reset during disabling of throttling?

Reply:
Yes I have captured task level throttledTasksCount. We are also planing to add this stats in stats API, it will help in that as well.

> As I understand this is a cumulative count, wondering should it get reset during disabling of throttling?

I don't see any reason for re-setting it to zero once we disable throttling. IMO we should maintain how many tasks were throttled in past. 
Do you have specific case in mind for which we need to reset it to zero?